qualtran/symbolics/types.py

```diff
@@ -26,6 +26,9 @@
 SymbolicComplex = Union[complex, sympy.Expr]
 document(SymbolicComplex, """A complex value or a sympy expression.""")
 
+AnySymbolic = TypeVar('AnySymbolic', SymbolicInt, SymbolicFloat)
+document(AnySymbolic, """A type variable that can be either SymbolicInt or SymbolicFloat.""")
```

Comment:
should this union in `SymbolicComplex` as well?

Reply:
why `TypeVar` instead of `Union` like the others?

Reply:
Let me try to give a technical explanation here --- I hope it's not too obvious or pedantic of one.

Assuming that mypy works along the lines of other type systems, the difference is that a `TypeVar` is a form of [bounded parametric polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism#Bounded_parametric_polymorphism) whereas a `Union` is a form of [width subtyping](https://en.wikipedia.org/wiki/Subtyping#Width_and_depth_subtyping).

What that mumbo jumbo means is that a function that accepts a `Union[int, float]` and returns a `Union[int, float]` is _one function_ whose more precise relationship between input and output (`int`s go to `int`s and `float`s go to `float`s) must be deduced through external means. 

In mypy, that external means is the `@overload` system, which unfortunately considers `Iterable[int | Expr]` as "the same or broader" parameter type as `Iterable[float | Expr]`, as shown in the playground. This behavior seems to be because `int` is implicitly convertible to `float` in Python, but I do not fully understand it. Whether it is a bug or feature I cannot tell, or whether there is yet some trick to get around the problem.

By contrast, a function that accepts `AnySymbolic` and returns `AnySymbolic` is a _family of functions_ parameterized by the type variable `AnySymbolic`. Here, we also _bound_ it to be either `SymbolicInt` or `SymbolicFloat`, but that is a totally ancillary thing. Formally, the type checker is obligated to treat `AnySymbolic` as if it were some totally opaque type `T`.

The rules of parametric polymorphism state that the type checker must check each function in this family independently, without considering any other. In particular, the fact that `int` is implicitly convertible to `float` cannot influence on the process that the type checker takes to determine that giving the function an `Iterable[int]` must cause it to output an `int`.

That is my best explanation so far of why `TypeVar` works here whereas `Union` does not.

Reply:
If it helps, `@overload` is analogous to function overloading in C++ whereas `TypeVar` is analogous to `template<typename T>`.

Reply:
thanks for the explanation!

Reply:
Yeah, we should name it `SymbolicT`, which is the convention for typevars elsewhere in the library and makes the distinction between the Union ones